optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/LogicalTuple.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Used inside Drools-based constraint stream implementation to implement grouping operations. This is the carrier data
+ * type which the grouping rule logically inserts into the Drools working memory, so that the following rules can work
+ * with the results of the grouping.
+ *
+ * Instances of this class are equal if they have the same {@link #getContext()} and their {@link #getItem(int)} all
+ * {@link Objects#equals(Object, Object)} as well. Drools will not re-insert a fact if equalling fact has already been
+ * inserted before.
+ */
+public final class LogicalTuple {
+
+    private final String context;
+    private final Object[] items;
+
+    public LogicalTuple(final String context, final Object... items) {
+        Objects.requireNonNull(context, "Logical tuple rule context must not be null.");
+        this.context = context;
+        int itemCount = items.length;
+        if (itemCount == 0) { // such a tuple makes no sense
+            throw new IllegalArgumentException("Logical tuple must have at least one element.");
+        } else if (itemCount > 3) { // there are no quad-streams yet
```

Comment:
If we do such a check (and more importantly, if there are similar check in other classes), it might be good to extract the value into some global constant named e.g. `SUPPORTED_CARDINALITY`.